blog/_src/posts/2017-05-12-artifacts-for-semantics.md

```diff
@@ -0,0 +1,155 @@
+indeed, the PhD PL course at Northeastern is entirely taught
+using [Redex](https://redex.racket-lang.org/), but this post is about a
+different approach -- writing small-step interepreters in OCaml and compiling
+them to run on the web -- which has its own advantages.
+
+<!-- more -->
+
+[Redex](https://redex.racket-lang.org/) is a wonderful tool for exploring
+programming language semantics, but rather than use Redex, we wrote a
+interpreter, typechecker, and parser in OCaml, compiled it to Javascript
+using [js_of_ocaml](http://ocsigen.org/js_of_ocaml/), and then put it on a
+webpage (with an editor with syntax highlighting and error reporting) that
+allows people to step through examples from the paper or write their own. While
+this probably involved more work on our part than Redex would have, we think it
+resulted in something that is more accessible to readers, and thus worth
+considering.
+
+This post will summarize the different parts to make it easier for others to
+repeat this process. We think it was a total success, and have gotten feedback
+that it makes understanding the (somewhat complex) language from the paper much
+easier. We argue that building such interpreters / typecheckers is easy enough
+that all papers should do this. Further, while our interpreter / typechecker is
+completely unverified, since we wrote it in OCaml, this approach should work
+equally well for semantics verified in Coq and then extracted to OCaml.
+
+------
+
+The paper in question,
+[FunTAL: Reasonably Mixing a Functional Language with Assembly](https://dbp.io/pubs/2017/funtal.pdf) (to
+appear in PLDI17), presents a multi-language that incorporates a typed assembly
+language and a simple functional language where each can be embedded within the
+other. The paper then develops a logical relation that can be used to reason
+about equivalence of such mixed programs. For example in the paper we show an
+imperative register-based factorial and a functional factorial equivalent.
+
+Both the static and dynamic semantics are relatively complex. The typed assembly
+has registers (which store word-sized values), a heap (which stores code-blocks
+and tuples), and a stack (not a call-stack, simply a place where word-sized
+values can be pushed and popped). Code-blocks have pre-conditions on the state
+of the registers and the stack, and allow the tail of the stack to be abstracted
+over polymorphically. This allows values to be protected on the stack before
+jumping to blocks that otherwise could change them. This is used, along with a
+novel notion of **return markers**, to ensure well-bracketing in the control
+flow of the typed assembly. The return markers indicate the location that points
+to the block that will eventually be returned to (assuming it doesn't loop
+infinitely). At the top level, the return marker `end` indicates that, assuming
+it does not loop, eventually the program will stop, rather than returning
+somewhere else. 
```

Comment:
off topic:
return markers sound like the _call-to-return_ edges that the IFDS framework adds to control-flow graphs, (https://research.cs.wisc.edu/wpis/papers/popl95.pdf Section 2) ... which I think they got from Sharir & Pneuli but I'm not sure